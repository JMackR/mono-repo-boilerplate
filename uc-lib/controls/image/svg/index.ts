export * from "./svg"
export * from "./svg.props-base"
