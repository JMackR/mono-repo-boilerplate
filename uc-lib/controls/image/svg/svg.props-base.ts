import { TextColors, Colors } from "uc-lib/themes"

export interface RemoteSVGSource {
  uri: string
  size: {
    width: number
    height: number
  }
}

/**
 * @warning !!!THIS SHLD NOT BE USED TSIDE OF THE SVG COMPONENT!!!
 * An extension on the Colors type to be used when passing color props to
 * the `SVG` member of LocalSVGSource.
 */
export interface SVGColors extends Colors {
  tintColor?: string
}

export interface LocalSVGSource {
  SVG: React.FunctionComponent<React.SVGProps<SVGSVGElement> & SVGColors>
  size?: {
    width: number | string
    height: number | string
  }
}

export interface SVGPropsBase {
  onPress?(): void
  tint?: keyof TextColors
  remoteSVG?: RemoteSVGSource
  localSVG?: LocalSVGSource
  /**
   * Used to locate this view in end-to-end tests.
   */
  testID?: string
}
