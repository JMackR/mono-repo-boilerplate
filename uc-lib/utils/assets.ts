import { PixelRatio } from "react-native"

const pixelDensityExtension = () => {
  const pixelRatio = PixelRatio.get()
  if (pixelRatio >= 3) {
    return "@3x"
  } else if (pixelRatio >= 2) {
    return "@2x"
  } else {
    return ""
  }
}

/**
 * Constructs and returns a formatted cdn url
 * @param path to asset including extension (e.g. "mojis/assassin-64.png")
 * @return the constructed cdn url
 */
export const cdnUrl = (path: string) => {
  const prefix = "https://wrkshp-cdn.com/com.nextmusic/"
  const url = `${prefix}${path}`
  const index = url.lastIndexOf(".")
  return url.slice(0, index) + pixelDensityExtension() + url.slice(index)
}
