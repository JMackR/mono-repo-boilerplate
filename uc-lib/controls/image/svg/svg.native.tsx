import React from "react"
import { SvgUri } from "react-native-svg"
import { SVGPropsNative } from "./svg.props.native"
import { useColor, useColorForTextColor } from "../../../themes"
import invariant from "invariant"
import { LocalSVGSource } from "./svg.props-base"
import { ClickableOpacity } from "../../clickable/clickable-opacity"
export const SVG = (props: SVGPropsNative) => {
  const { remoteSVG, localSVG, tint, testID } = props
  invariant(remoteSVG || localSVG, "'localSVG' or 'remoteSVG' are required for SVG to operate correctly")

  const { colors } = useColor()
  const iconTintColor = useColorForTextColor(tint!)
  if (remoteSVG) {
    invariant(tint === undefined, 'The property "tint" is not supported for remoteSVGs: ' + JSON.stringify(tint))
    const { size, uri } = remoteSVG
    return <SvgUri {...size} uri={uri} onPress={props.onPress} />
  } else {
    const { SVG, size } = localSVG as LocalSVGSource
    // WARNING: Do not assign testID to SVG directly
    if (props.onPress) {
      return (
        <ClickableOpacity onClick={props.onPress} testID={testID}>
          <SVG {...size} {...colors} tintColor={iconTintColor} />
        </ClickableOpacity>
      )
    } else {
      return <SVG {...size} {...colors} tintColor={iconTintColor} />
    }
  }
}
