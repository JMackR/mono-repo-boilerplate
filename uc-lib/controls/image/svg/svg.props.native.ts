import { SVGPropsBase } from "./svg.props-base"

// tslint:disable-next-line: no-empty-interface
export interface SVGPropsNative extends SVGPropsBase {}
