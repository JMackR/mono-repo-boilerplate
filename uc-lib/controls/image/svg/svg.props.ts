import { SVGPropsBase } from "./svg.props-base"

export interface SVGPropsWeb extends SVGPropsBase {
  className?: string
}
