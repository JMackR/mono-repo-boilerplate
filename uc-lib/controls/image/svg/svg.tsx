import React from "react"
import { SVGPropsWeb } from "./svg.props"
import { useColor, useColorForTextColor, Colors } from "uc-lib/themes"
import invariant from "invariant"

interface IndexableColors extends Colors {
  [key: string]: string
}

const convertObjectKeysToLowercase = (obj: IndexableColors) => {
  const keys = Object.keys(obj)
  let n = keys.length
  const convertedObject: { [key: string]: string } = {}
  while (n--) {
    const key = keys[n]
    convertedObject[key.toLowerCase()] = obj[key]
  }
  return convertedObject
}

export const SVG = (props: SVGPropsWeb) => {
  const { remoteSVG, localSVG, tint } = props
  invariant(remoteSVG || localSVG, "'localSVG' or 'remoteSVG' are required for SVG to operate correctly")

  const { colors } = useColor()
  const iconTintColor = useColorForTextColor(tint!)

  if (remoteSVG) {
    invariant(tint === undefined, 'The property "tint" is not supported for remoteSVGs: ' + JSON.stringify(tint))
    const { size, uri } = remoteSVG
    return <img src={uri} style={{ ...size }} onClick={props.onPress} />
  } else {
    // React and React-Native diverge in how they handle props be forwarded to potential DOM elements.
    // Below, we are forwarding all of these colors to the SVG tag to use their dynamic values.
    // React.js requires that these props all be lowercase.
    // https://reactjs.org/warnings/unknown-prop.html
    const convertedColors = convertObjectKeysToLowercase(colors as IndexableColors) as IndexableColors
    convertedColors.tintcolor = iconTintColor

    const { SVG, size } = localSVG!
    return <SVG {...size} {...convertedColors} className={props.className} onClick={props.onPress} />
  }
}
