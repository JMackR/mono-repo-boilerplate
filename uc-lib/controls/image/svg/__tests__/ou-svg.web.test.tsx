import React from "react"
import * as renderer from "react-test-renderer"
import { shallow } from "enzyme"
import { SVG } from "../svg"
import { RemoteSVGSource } from "../svg.props-base"
import { TabBarHomeFill } from "../../../../assets"

describe("SVG Snapshot Tests", () => {
  const remoteSVGData: RemoteSVGSource = {
    uri: "https://dev.w3.org/SVG/tools/svgweb/samples/svg-files/compass.svg",
    size: {
      width: 120,
      height: 120,
    },
  }

  test("SVG renders remoteSVG", () => {
    const tree = renderer.create(<SVG remoteSVG={remoteSVGData} />)
    expect(tree).toMatchSnapshot()
  })

  test("SVG does not render remoteSVG with tint parameter", () => {
    expect(() => shallow(<SVG remoteSVG={remoteSVGData} tint="error" />)).toThrowError()
  })

  test("SVG renders tintable localSVG with error tint", () => {
    const tree = renderer.create(<SVG localSVG={TabBarHomeFill} tint="error" />)
    expect(tree).toMatchSnapshot()
  })

  test("SVG renders tintable localSVG with link tint", () => {
    const tree = renderer.create(<SVG localSVG={TabBarHomeFill} tint="brand" />)
    expect(tree).toMatchSnapshot()
  })

  test("SVG renders tintable localSVG without tint", () => {
    const tree = renderer.create(<SVG localSVG={TabBarHomeFill} />)
    expect(tree).toMatchSnapshot()
  })

  test("SVG does not render with no SVG parameter", () => {
    expect(() => shallow(<SVG />)).toThrowError()
  })
})
