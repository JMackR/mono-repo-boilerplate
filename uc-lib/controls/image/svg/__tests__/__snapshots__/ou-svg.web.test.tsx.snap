// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVG Snapshot Tests SVG renders localSVG  1`] = `
<tru-ycircle-fill.svg
  alwaysdark="#121212"
  alwayslight="#FFFFFF"
  basalt="#8A8A8A"
  crystal="#FFFFFF"
  crystalhover="#F7F7F7"
  crystalpressed="#EEEEEE"
  disabled="#FAFAFA"
  glacialblue="#3D98D2"
  granite="#ABABAB"
  larchyellow="#F4BE19"
  larchyellowhover="#F5D624"
  larchyellowpressed="#F3A70C"
  limestone="#E2E2E2"
  limestonehover="#EEEEEE"
  limestonepressed="#D0D0D0"
  obsidian="#121212"
  overlay="rgba(18,18,18,0.3)"
  paintbrushred="#E05666"
  palousegreen="#00A87E"
  palousegreenhighlight="#DEF3EC"
  palousegreenhover="#43B893"
  palousegreenpressed="#00986A"
  quartz="#FAFAFA"
/>
`;

exports[`SVG Snapshot Tests SVG renders localSVG with overwritten width/height 1`] = `
<tru-ycircle-fill.svg
  alwaysdark="#121212"
  alwayslight="#FFFFFF"
  basalt="#8A8A8A"
  crystal="#FFFFFF"
  crystalhover="#F7F7F7"
  crystalpressed="#EEEEEE"
  disabled="#FAFAFA"
  glacialblue="#3D98D2"
  granite="#ABABAB"
  height={120}
  larchyellow="#F4BE19"
  larchyellowhover="#F5D624"
  larchyellowpressed="#F3A70C"
  limestone="#E2E2E2"
  limestonehover="#EEEEEE"
  limestonepressed="#D0D0D0"
  obsidian="#121212"
  overlay="rgba(18,18,18,0.3)"
  paintbrushred="#E05666"
  palousegreen="#00A87E"
  palousegreenhighlight="#DEF3EC"
  palousegreenhover="#43B893"
  palousegreenpressed="#00986A"
  quartz="#FAFAFA"
  width={120}
/>
`;

exports[`SVG Snapshot Tests SVG renders remoteSVG 1`] = `
<img
  src="https://dev.w3.org/SVG/tools/svgweb/samples/svg-files/compass.svg"
  style={
    Object {
      "height": 120,
      "width": 120,
    }
  }
/>
`;

exports[`SVG Snapshot Tests SVG renders tintable localSVG with error tint 1`] = `
<home-fill.svg
  alwaysdark="#121212"
  alwayslight="#FFFFFF"
  basalt="#8A8A8A"
  crystal="#FFFFFF"
  crystalhover="#F7F7F7"
  crystalpressed="#EEEEEE"
  disabled="#FAFAFA"
  glacialblue="#3D98D2"
  granite="#ABABAB"
  larchyellow="#F4BE19"
  larchyellowhover="#F5D624"
  larchyellowpressed="#F3A70C"
  limestone="#E2E2E2"
  limestonehover="#EEEEEE"
  limestonepressed="#D0D0D0"
  obsidian="#121212"
  overlay="rgba(18,18,18,0.3)"
  paintbrushred="#E05666"
  palousegreen="#00A87E"
  palousegreenhighlight="#DEF3EC"
  palousegreenhover="#43B893"
  palousegreenpressed="#00986A"
  quartz="#FAFAFA"
  tintcolor="#E05666"
/>
`;

exports[`SVG Snapshot Tests SVG renders tintable localSVG with link tint 1`] = `
<home-fill.svg
  alwaysdark="#121212"
  alwayslight="#FFFFFF"
  basalt="#8A8A8A"
  crystal="#FFFFFF"
  crystalhover="#F7F7F7"
  crystalpressed="#EEEEEE"
  disabled="#FAFAFA"
  glacialblue="#3D98D2"
  granite="#ABABAB"
  larchyellow="#F4BE19"
  larchyellowhover="#F5D624"
  larchyellowpressed="#F3A70C"
  limestone="#E2E2E2"
  limestonehover="#EEEEEE"
  limestonepressed="#D0D0D0"
  obsidian="#121212"
  overlay="rgba(18,18,18,0.3)"
  paintbrushred="#E05666"
  palousegreen="#00A87E"
  palousegreenhighlight="#DEF3EC"
  palousegreenhover="#43B893"
  palousegreenpressed="#00986A"
  quartz="#FAFAFA"
  tintcolor="#00A87E"
/>
`;

exports[`SVG Snapshot Tests SVG renders tintable localSVG without tint 1`] = `
<home-fill.svg
  alwaysdark="#121212"
  alwayslight="#FFFFFF"
  basalt="#8A8A8A"
  crystal="#FFFFFF"
  crystalhover="#F7F7F7"
  crystalpressed="#EEEEEE"
  disabled="#FAFAFA"
  glacialblue="#3D98D2"
  granite="#ABABAB"
  larchyellow="#F4BE19"
  larchyellowhover="#F5D624"
  larchyellowpressed="#F3A70C"
  limestone="#E2E2E2"
  limestonehover="#EEEEEE"
  limestonepressed="#D0D0D0"
  obsidian="#121212"
  overlay="rgba(18,18,18,0.3)"
  paintbrushred="#E05666"
  palousegreen="#00A87E"
  palousegreenhighlight="#DEF3EC"
  palousegreenhover="#43B893"
  palousegreenpressed="#00986A"
  quartz="#FAFAFA"
/>
`;
